### PR TITLE
Support python 3.11 and aiohttp 3.8

### PR DESCRIPTION
**Pull request by marcinsulikowski**
### Remove remnants of aiohttp 2.x and Python 3.5 support

Commit adfb65038caebce3d62 removed support for aiohttp 2.x and Python
3.5 but there are still some remnants of these left. We get rid of them.
This also removes one `mypy` warning:

> compat.py:31: error: All conditional function variants must have identical signatures [misc]

### Remove aiohttp 2.x from tox.ini

We no longer run tests against aiohttp 2.x so we don't need it in
`tox.ini`.

### Add Python 3.10 support

aioresponses works with Python 3.10 with no issues so we update the
package metadata to say that, and we update the list of test
environments to include Python 3.10.

### Add tests against aiohttp 3.8

aioresponses works with aiohttp 3.8 so we extend the list of test
environments to include this version.

### Drop Python 3.6 support

We cannot run tests with Python 3.10 using Pytest 6.x due to [1]. This
is fixed in Pytest 7.x but Pytest 7.x no longer supports Python 3.6
which aioresponses tries to support. However because Python 3.6 is
already past the EOL date, the simplest solution is to no longer test
with Python 3.6 and to stop declaring support for that version.

[1] https://github.com/pytest-dev/pytest/issues/8546

### Update Pytest to the latest version to fix tests on Python 3.10

Our tests are filing on Python 3.10 due to [1] so we update Pytest in
`requirements-dev.txt` to the latest version.

[1] https://github.com/pytest-dev/pytest/issues/8546

### Don't test with aiohttp 3.6 and older on Python 3.10

aiohttp 3.6 and older use Python features which were removed in Python
3.10:

                assert port is not None
    >           hosts = await asyncio.shield(self._resolve_host(
                    host,
                    port,
                    traces=traces), loop=self._loop)
    E               TypeError: shield() got an unexpected keyword argument 'loop'

    .tox/py3.10-aiohttp35/lib/python3.10/site-packages/aiohttp/connector.py:952: TypeError

To avoid test failures, we remove unsupported combinations from GitHub
workflows.

### Drop support for aiohttp 3.2.x and older

aiohttp 3.2.x and older don't work on Python 3.7 and because we now
require Python 3.7+, we can drop support for aiohttp 3.0, 3.1, and 3.2.

### Add Python 3.11 support

aioresponses works with Python 3.11 with no issues so we update the
package metadata to say that, and we update the list of test
environments to include Python 3.11. Python 3.11 is supported by aiohttp
starting from its v3.8 ([1]), and we also need newer `yarl` ([2]).

[1] https://github.com/aio-libs/aiohttp/issues/6600
[2] https://github.com/aio-libs/yarl/issues/706

### CODE REVIEW

**Comment by marcinsulikowski** [comment]
A couple places in README files should also be updated but this is covered by #223

**Comment by marcinsulikowski** [comment]
I needed a few attempts but it seems that all test combinations are finally passing.

**Comment by marcinsulikowski** [comment]
But the coverage report tells that there's some dead code. Let me fix that as well.

**Comment by pnuckowski** [comment]
Wow, a lot of work was done :)

**Comment by gaby** [comment]
Looking forward to seeing this get merged! 🎉

**Comment by marcinsulikowski** [comment]
Things changed a but since I created this PR. I will add conflict resolution and support for 3.11.

**Comment by gaby** [comment]
@pnuckowski This PR is ready now

**Comment by avalanchy** [comment]
@pnuckowski bump!

File: aioresponses/compat.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 from urllib.parse import parse_qsl, urlencode
 
 from aiohttp import __version__ as aiohttp_version, StreamReader
+from aiohttp.client_proto import ResponseHandler
 from multidict import MultiDict
 from pkg_resources import parse_version
```
Comment (iamibi) [inline aioresponses/compat.py:10]: The package throws the following warning when running locally
```
DeprecationWarning: pkg_resources is deprecated as an API
```
The official documentation suggests the use of [importlib.resources](https://docs.python.org/3/library/importlib.resources.html#module-importlib.resources), [importlib.metadata](https://docs.python.org/3/library/importlib.metadata.html#module-importlib.metadata) as a replacement. More info here: https://setuptools.pypa.io/en/stable/pkg_resources.html

Comment (gaby) [reply aioresponses/compat.py:10]: @brettdh I'm wondering the same thing :-S

Comment (brettdh) [reply aioresponses/compat.py:10]: I can confirm that all unit tests pass in Python 3.11.4 after changing this line (and one other like it, in `tests/test_aioresponses.py`) to this:

```python
from packaging.version import parse as parse_version
```

Using a private fork for now, but hoping that can get fixed, merged, and released soon.

**Review by pnuckowski — APPROVED** [review]

**Comment by gaby** [comment]
@pnuckowski Thank you! 💪

**Comment by brettdh** [comment]
I was very confused about:
1. why this was merged without the `pkg_resources` fix above; and
2. why [tests are passing on 3.11](https://github.com/pnuckowski/aioresponses/actions/runs/5316173825/jobs/9625410529) even without that fix; and
3. why `import aioresponses` still failed in my project's 3.11 virtualenv; AND
4. why `import pkg_resources` failed in my project's 3.11 virtualenv but succeeded in my aioresponses 3.11 virtualenv

and then I realised that my aioresponses virtualenv also includes `setuptools`, and my other project's virtualenv does not (for some reason, perhaps owing to me having created the former with `python -m venv` and the other with `pyenv virtualenv create`). 😑 Mystery (mostly) solved.

@pnuckowski Thanks for the quick response and for merging! Looking forward to a new release.

**Comment by brettdh** [comment]
Er, whoops; celebrated too soon. The `pkg_resources` issue above means that aioresponses still fails with this error in my project's unit tests (under `pytest`):

```
ImportError while importing test module '/Users/brhiggins/src/myproj/tests/test_foo.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../../../.pyenv/versions/3.11.4/lib/python3.11/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_foo.py:6: in <module>
    from aioresponses import aioresponses
../../../../.pyenv/versions/3.11.4/envs/myproj/lib/python3.11/site-packages/aioresponses/__init__.py:2: in <module>
    from .core import CallbackResult, aioresponses
../../../../.pyenv/versions/3.11.4/envs/myproj/lib/python3.11/site-packages/aioresponses/core.py:33: in <module>
    from .compat import (
../../../../.pyenv/versions/3.11.4/envs/myproj/lib/python3.11/site-packages/aioresponses/compat.py:10: in <module>
    from pkg_resources import parse_version
E   ModuleNotFoundError: No module named 'pkg_resources'
```

This is with aioresponses installed from `56b843319d5d0ae8a405f188e68d7ba8c7573bc8`.

**Comment by iamibi** [comment]
> Er, whoops; celebrated too soon. The `pkg_resources` issue above means that aioresponses still fails with this error in my project's unit tests (under `pytest`):
> 
> ```
> ImportError while importing test module '/Users/brhiggins/src/myproj/tests/test_foo.py'.
> Hint: make sure your test modules/packages have valid Python names.
> Traceback:
> ../../../../.pyenv/versions/3.11.4/lib/python3.11/importlib/__init__.py:126: in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
> tests/test_foo.py:6: in <module>
>     from aioresponses import aioresponses
> ../../../../.pyenv/versions/3.11.4/envs/myproj/lib/python3.11/site-packages/aioresponses/__init__.py:2: in <module>
>     from .core import CallbackResult, aioresponses
> ../../../../.pyenv/versions/3.11.4/envs/myproj/lib/python3.11/site-packages/aioresponses/core.py:33: in <module>
>     from .compat import (
> ../../../../.pyenv/versions/3.11.4/envs/myproj/lib/python3.11/site-packages/aioresponses/compat.py:10: in <module>
>     from pkg_resources import parse_version
> E   ModuleNotFoundError: No module named 'pkg_resources'
> ```
> 
> This is with aioresponses installed from `56b843319d5d0ae8a405f188e68d7ba8c7573bc8`.

The workaround is to install the `setuptools` when running test cases. Specially in case of CI builds. Any latest version of setuptools should work.

**Comment by brettdh** [comment]
@iamibi Thanks; that works for now.

**Pull request merged**